pinot-timeseries/pinot-timeseries-spi/src/main/java/org/apache/pinot/tsdb/spi/series/BaseTimeSeriesBuilder.java

```diff
@@ -81,4 +84,9 @@ public void mergeAlignedSeriesBuilder(BaseTimeSeriesBuilder builder) {
   }
 
   public abstract TimeSeries build();
+
+  /**
+   * Used by the leaf stage, because the leaf stage materializes tag values very late.
+   */
+  public abstract TimeSeries buildWithTagOverrides(List<String> tagNames, Object[] tagValues);
```

Comment:
@raghavyadav01 : this one change will be required in series builders. This should do the build, AND use the provided tag names and values for the new series. See other builders in this PR for reference.